File: dev/requirements/logic/cases/started-count.4.7.case.swift
import Foundation
import SetlistCore

/// 4.7 — The set-list header counts jokes started out of jokes planned,
/// counting a re-opened joke once.
let startedCount_4_7 = RequirementCase(id: "4.7") {
    let jokes = (1 ... 9).map { Joke(title: "Bit \($0)", body: "…") }
    var stage = StageModel(jokes: jokes, plannedLength: 1200)

    try expectEqual(stage.startedCount, 0, "nothing started yet")
    try expectEqual(stage.plannedCount, 9, "jokes planned")

    stage.tapCard(jokeID: jokes[0].id, at: CaptureTime(0))
    stage.tapCard(jokeID: jokes[1].id, at: CaptureTime(190))
    stage.tapCard(jokeID: jokes[2].id, at: CaptureTime(315))
    stage.tapCard(jokeID: jokes[3].id, at: CaptureTime(423))
    try expectEqual(stage.startedCount, 4, "four of nine started")

    // Going back to an earlier bit does not make it a fifth.
    stage.tapCard(jokeID: jokes[1].id, at: CaptureTime(507))
    try expectEqual(stage.startedCount, 4, "a re-opened joke counts once")
}

File: dev/requirements/logic/cases/started-count.4.7.case.dart
import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 4.7 — A joke re-opened after being told still counts once in the started
/// tally.
final startedCount_4_7 = RequirementCase('4.7', (tester) {
  final stage = StageModel(
    jokes: [
      Joke(id: 'opener', title: 'Opener', body: ''),
      Joke(id: 'van', title: 'Van', body: ''),
      Joke(id: 'closer', title: 'Closer', body: ''),
    ],
    plannedLength: const Duration(minutes: 20),
  );

  expectThat(stage.startedCount, 0, 'nothing is started before anything is tapped');

  stage.tapCard(jokeId: 'van', at: const CaptureTime(0));
  stage.tapCard(jokeId: 'opener', at: const CaptureTime(60));
  stage.tapCard(jokeId: 'van', at: const CaptureTime(200));

  expectThat(stage.segments.length, 3, 'the van has two segments');
  expectThat(stage.startedCount, 2, 'but counts once in the tally');
  expectThat(stage.plannedCount, 3, 'out of the jokes planned');
});
